### PR TITLE
Upgrade the unikernel to git.3.1.0

### PR DESCRIPTION
**Pull request by dinosaure**
This PR needs ocaml/opam-repository#17966 (or `git.3.1.0`). It updates:
- the `config.ml` and import `mimic`'s stuffs 
  this part is a bit cryptic. The idea is to _fill_ a `Mimic.ctx` according to user's arguments (such as `ssh-seed`, `ssh-auth` and `remote`). This way to play with `mimic` is done without any DNS resolvers - even if we are able to add one (or add a `Mimic.fold : [ host ] Domain_name.t -> Ipaddr.V4.t`). With that, we are able to be synchronize to a local Git repository (`git://`, `git@...`, `http://`).
- An upgrade to `irmin.2.3.0` which shadows `Store.tree`

Some notes:
- [ ] we don't support IPv6 yet but if the server can use it, I can upgrade the rest to IPv6 easily
- [x] `https` is not available due to the remove of `conduit`/`resolver` - in such context and specifically for this unikernel, we should not have a DNS resolver which is needed for `https` in anyway
- [x] remove `mimic_conf` which is not used (it's a merge operation from 2 `Mimic.ctx` with `functoria`)

If you have any question, let me know :+1:. I can go deeper about details (and `mimic`).

### CODE REVIEW

**Comment by dinosaure** [comment]
Ah and currently `x25519.net` uses this version and it seems fine.

**Comment by hannesm** [comment]
I'll need some time to review this change. It would be easier for me if this would be based on the ssh branch of this repository (see #1).

**Comment by dinosaure** [comment]
It will be hard to follow the same trunk from #1 when:
1) we don't use `conduit.2` anymore
2) we use `mimic` instead which works differently (I mean, the `with_ssh` from #1 is located to the `config.ml` now)
3) the way to represent the `remote` was updated from an `Uri.t` (with headers...) to a `Smart_git.Endpoint.t`

**Comment by hannesm** [comment]
ok, thanks. CI failed due to:

```
+ ocamlfind ocamldep -package 'checkseum.c digestif.c' -package tcpip.udp -package tcpip.tcp -package tcpip.stack-direct -package tcpip.ipv4 -package tcpip.icmpv4 -package tcpip -package mirage-types-lwt -package mirage-types -package mirage-solo5 -package mirage-runtime -package mirage-net-solo5 -package mirage-logs -package mirage-crypto-rng-mirage -package mirage-clock-freestanding -package mirage-bootvar-solo5 -package lwt -package logs -package irmin-mirage-git -package irmin-mirage -package git-mirage.tcp -package git-mirage.ssh -package git-mirage -package git -package functoria-runtime -package ethernet -package dns-tsig -package dns-server.zone -package dns-server.mirage -package dns-server -package conduit-mirage -package arp.mirage -package arp -modules main.ml > main.ml.depends
ocamlfind: Package `ppx_deriving' not found - required by `ppx_irmin'
Command exited with code 2.
```

is this known (somewhere a missing dependency)? what is using ppx_irmin?

**Comment by dinosaure** [comment]
The error seems to come from `irmin` and `ppx_irmin`, currently we don't know why but @CraigFe will take a look on that :+1:. Currently, you just need to install `ppx_deriving` and the error does not occur.

**Comment by CraigFe** [comment]
After some reading, it seems likely that this is a consequence of https://github.com/ocaml/dune/issues/1327; `dune` adding an implicit dependency on `ppx_deriving` for use with "other build systems".

It seems that the intended solution is to add an explicit opam dependency `ppx_irmin` → `ppx_deriving`, but I'd appreciate a thumbs up from someone with better knowledge before I go ahead with that change. I'm interested in knowing why [`ppx_sexp_conv`](https://github.com/janestreet/ppx_sexp_conv/blob/master/ppx_sexp_conv.opam) doesn't have a `ppx_deriving` dependency, for instance. Will ask around a bit.

**Comment by hannesm** [comment]
Thanks for investigating this issue - I'm by no means an ppx / ppx_deriving expert, but fixing the dependencies in opam-repository would be good (even better if a CI would catch them).

**Comment by hannesm** [comment]
The ppx_sexp_conv story as far as I remember is that its build-time only, and runtime there's a dependency on ppx_sexp_conv.runtime-lib (which does not include the ppx driver etc.) -- for irmin / ppx_irmin I'm not sure how it looks like, but eventually (in irmin's dune file) you'll need to revise "(libraries ppx_irmin.runtime-lib)" and provide such a runtime-lib sublibrary -- to avoid having to link compiler-libs and ppx stuff into a unikernel. Unfortunately I fail to remember the details and who's adding the dependency on ppx_sexp_conv.runtime-lib (but from their META file: `requires(-ppx_driver) = "ppx_sexp_conv.runtime-lib"`.

**Comment by dinosaure** [comment]
It should be fixed now, can you restart the CI?

**Comment by hannesm** [comment]
ok, thanks for notifying. what about this ppx_repr 0.2.0 package in opam-repository? it is installable and lacks the ppx_deriving dependency, thus will lead to a broken system -- any plans to either add the dependency or mark it unavailable?

**Comment by hannesm** [comment]
thanks, the CI is green now. I'll review and merge this week.

**Comment by dinosaure** [comment]
I upgraded the unikernel with `git.3.2.0`. One feature remains: how to push over https? Indeed, in that case, we must pass a token to be allowed to push to a Git repository. The ability to make an unikernel without a DNS server still depends on mirage/mirage#1210 but I think the PR is ready.

**Comment by hannesm** [comment]
Thanks @dinosaure 

> The ability to make an unikernel without a DNS server

You meant a unikernel without a DNS resolver (client)?

> One feature remains: how to push over https?

I'd welcome to be able to use `git+https://user:password@host/git-repo` as a remote (i.e. encoding the authentication information into the url).

**Comment by dinosaure** [comment]
> You meant a unikernel without a DNS resolver (client)?

Yes!

> I'd welcome to be able to use git+https://user:password@host/git-repo as a remote (i.e. encoding the authentication information into the url).

It should be possible to do that indeed. `Smart_git.Endpoint.t` keeps HTTP  headers on its side so I think we should improve how to parse the given string and fill the `Smart_git.Endpoint.t` which these informations. I need to check how GitHub works when we wants to push on a repository.

**Comment by dinosaure** [comment]
According `ocaml-git.3.3.0`, we support `fetch`/`push` over HTTPS as you requested with `-r https://user:password@github.com/repo` :+1:.

**Comment by hannesm** [comment]
thanks @dinosaure, I have minor tweaks and opened #3 for this purpose (namely keep the commit from #1 in the history, and retain the boot arguments). I'll close this PR, but your work (commit) is preserved in #3 which I'll merge as soon as CI is back.
